java-checks/src/main/java/org/sonar/java/checks/regex/PossessiveQuantifierContinuationCheck.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.regex;
+
+import java.util.Collections;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.helpers.RegexTreeHelper;
+import org.sonar.java.checks.helpers.SubAutomaton;
+import org.sonar.java.regex.RegexCheck;
+import org.sonar.java.regex.RegexParseResult;
+import org.sonar.java.regex.ast.AutomatonState;
+import org.sonar.java.regex.ast.FinalState;
+import org.sonar.java.regex.ast.Quantifier;
+import org.sonar.java.regex.ast.RegexBaseVisitor;
+import org.sonar.java.regex.ast.RegexSyntaxElement;
+import org.sonar.java.regex.ast.RepetitionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+
+@Rule(key = "S5994")
+public class PossessiveQuantifierContinuationCheck extends AbstractRegexCheck {
+
+  private static final String MESSAGE = "Change this impossible to match sub-pattern that conflicts with the previous possessive quantifier.";
+
+  @Override
+  public void checkRegex(RegexParseResult regexForLiterals, MethodInvocationTree mit) {
+    new Visitor(regexForLiterals.getFinalState()).visit(regexForLiterals);
+  }
+
+  private class Visitor extends RegexBaseVisitor {
+
+    private final FinalState finalState;
+
+    public Visitor(FinalState finalState) {
+      this.finalState = finalState;
+    }
+
+    @Override
+    public void visitRepetition(RepetitionTree repetitionTree) {
+      AutomatonState continuation = repetitionTree.continuation();
+      if (doesRepetitionContinuationAlwaysFail(repetitionTree) && continuation instanceof RegexSyntaxElement) {
```

Comment:
I believe we should still report the issue if the continuation is not a RegexSyntaxElement. If the non-regex-syntax continuation has a single successor (i.e. it's not a branch), we can just report on that successor instead (or, if it's still not a regex syntax element, keep following single edges until we find one or a branch). And if it's a branch (as might be the case for something like `(xx++)+x`) we should probably still report by only reporting on the repetition itself with the message adjusted accordingly.